examples/Superresolution/enet-pat.py

```diff
@@ -0,0 +1,334 @@
+                                BATCH_SIZE:BATCH_SIZE * 2, i:i + p, j:j + p, :]
+
+                            current_patch_a = gram_matrix(current_patch_a)
+                            current_patch_b = gram_matrix(current_patch_b)
+
+                            rsl.append(mse(current_patch_a, current_patch_b))
+                    # rsl is list of [b, p, p, c]
+                    return tf.add_n(rsl)
+
+                texture_loss_conv1_1 = tf.identity(texture_loss(conv1_1), name='normalized_conv1_1')
+                texture_loss_conv2_1 = tf.identity(texture_loss(conv2_1), name='normalized_conv2_1')
+                texture_loss_conv3_1 = tf.identity(texture_loss(conv3_1), name='normalized_conv3_1')
+
+            return [pool2_loss, pool5_loss, texture_loss_conv1_1, texture_loss_conv2_1, texture_loss_conv3_1]
+
+        with tf.variable_scope('gen'):
+            fake_hr = generator(Ilr, Ibicubic)
+            real_hr = Ihr
+
+        tf.identity(add_VGGMeans(fake_hr), name='prediction')
+
+        if not self.inference:
+            with tf.variable_scope('discrim'):
+                real_score = discriminator(real_hr)
+                fake_score = discriminator(fake_hr)
+
+            self.build_losses(real_score, fake_score)
+
+            if ctx.is_training:
+                additional_losses = additional_losses(fake_hr, real_hr)
+
+                with tf.name_scope('additional_losses'):
+                    # see table 2 from appendix
+                    loss = []
+                    loss.append(tf.multiply(1., self.g_loss, name="loss_LA"))
+                    loss.append(tf.multiply(2e-1, additional_losses[0], name="loss_LP1"))
+                    loss.append(tf.multiply(2e-2, additional_losses[1], name="loss_LP2"))
+                    loss.append(tf.multiply(3e-7, additional_losses[2], name="loss_LT1"))
+                    loss.append(tf.multiply(1e-6, additional_losses[3], name="loss_LT2"))
+                    loss.append(tf.multiply(1e-6, additional_losses[4], name="loss_LT3"))
+
+                self.g_loss = self.g_loss + tf.add_n(loss, name='total_g_loss')
+                add_moving_summary(self.g_loss, *loss)
+
+            fake_hr = add_VGGMeans(fake_hr)
+            real_hr = add_VGGMeans(real_hr)
+            Ibicubic = add_VGGMeans(Ibicubic)
+
+            if ctx.is_training:
+                # visualization
+                viz = (tf.concat([Ibicubic, fake_hr, real_hr], 2)) * 255.
+                viz = tf.cast(tf.clip_by_value(viz, 0, 255), tf.uint8, name='viz')
+                tf.summary.image('input,fake,real', viz,
+                                 max_outputs=max(30, BATCH_SIZE))
+
+            self.collect_variables()
+
+    def _get_optimizer(self):
+        lr = tf.get_variable(
+            'learning_rate', initializer=1e-4, trainable=False)
+        opt = tf.train.AdamOptimizer(lr)
+        gradprocs = [gradproc.ScaleGradient([('VGG19/*', 0.)]), gradproc.ScaleGradient([('discrim/*', 0.3)])]
```

Comment:
use tf.stop_gradient for VGG?
Even if lr=0 gradients are still computed and wasted

Reply:
any other solution?

Reply:
`with varreplace.freeze_variables():`

Reply:
Does this affects the gradients wrt. to `fake_hr`?

Reply:
No it'll only stops the gradients on variables (inside the with block), not on any layer output.

Reply:
That's awesome!

Reply:
fixed